### PR TITLE
Provide feature to compare repo with specific branch and tag.

### PR DESCRIPTION
**Pull request by rakshithtb**
Provide feature to compare repo with specific branch and tag.

### CODE REVIEW

File: src/Package/ComparePackageRepo.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Yaml Difference Report.
+     */
+    public function yamlDiffReport() {
+        $create_directory = "mkdir -p $this->comparisionFolder";
+        shell_exec($create_directory);
+
+        $this->accessToken = trim(readline('Please Enter Private Access token for Package Management Repo'));
+        $list_of_projects_cmd = "curl --header 'PRIVATE-TOKEN: $this->accessToken' '$this->gitlabApiPath/groups/$this->packageRepoGroup/projects?per_page=500' >> ./$this->comparisionFolder/$this->projectsPath";
```
Comment (garrowsmith) [inline src/Package/ComparePackageRepo.php]: @rakshithtb this is hard wired to Gitlabs, I'm going to merge to unblock testing, but create an Issue to abstract this as the library should allow other sources e.g. github, OR namespace the functions so it's clear these classes only are intended to support Gitlabs.

File: src/Package/ConfigInitializer.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     * @var string
+     */
+    protected $comparisionFolder = 'package-comparision';
+
+    /**
+     * Projects Details.
+     *
+     * @var string
+     */
+    protected $projectsPath = "projects.json";
```
Comment (garrowsmith) [inline src/Package/ConfigInitializer.php]: @rakshithtb nitpick - I'd provide more descriptive information - what is this file, is it autogenerated?

File: src/Package/ConfigInitializer.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Package Repo Group.
+     *
+     * @var string
+     */
+    protected $packageRepoGroup;
+
+    /**
+     * Comparision Folder.
```
Comment (garrowsmith) [inline src/Package/ConfigInitializer.php]: @rakshithtb typo... Comparision => Comparison. Please fix all instances :(

File: src/Package/ComparePackageRepo.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    public function __construct() {
+        parent::__construct();
+        $this->yamlDiffReport();
+    }
+
+    /**
+     * Yaml Difference Report.
+     */
+    public function yamlDiffReport() {
```
Comment (garrowsmith) [inline src/Package/ComparePackageRepo.php]: @rakshithtb recommend splitting this up into logical functions as rationale is this allows us to more easily unit tests different stages, provide mocks. We want to allow test coverage due to the role this plays in the development process.

e.g.

yamlDiffReport()
* Generating list of projects
* Download report
* Invoke comparison logic

**Review by garrowsmith — CHANGES_REQUESTED** [review]
Provided review comments, would like some addition to Readme.md or a PackageRepo.md to detail any setup or examples.

File: src/Package/GitLab/ConfigInitializer.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+   * @var string
+   */
+  protected $tarBranchTag;
+
+  /**
+   * GitLab API Path.
+   *
+   * @var string
+   */
+  protected $gitlabApiPath = "https://gitlab.bayer.com/api/v4";
```
Comment (guptahemant) [inline src/Package/GitLab/ConfigInitializer.php:40]: This should be configurable, and not hard coded to bayer gitlab.

File: src/Package/GitLab/ComparePackageRepo.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  private function getProjectsList() {
+    $project_details = [];
+
+    // Remove old Entries.
+    if (is_dir($this->comparisonFolder)) {
+      $remove_directory = "rm -r $this->comparisonFolder";
+      shell_exec($remove_directory);
+    }
+
+    // Create comparision folder.
```
Comment (garrowsmith) [inline src/Package/GitLab/ComparePackageRepo.php]: Nitpick, one 'comparision' typo left, but I'm not blocking the MR @rakshithtb.

**Comment by rakshithtb** [comment]
Closing this MR

**Pull request closed without merging**